server/src/main/java/io/unitycatalog/server/service/FunctionService.java

```diff
@@ -2,43 +2,159 @@
+    this.authorizer = authorizer;
+    evaluator = new UnityAccessEvaluator(authorizer);
+  }
 
   @Post("")
-  public HttpResponse createFunction(CreateFunctionRequest createFunctionRequest) {
-    return HttpResponse.ofJson(FUNCTION_REPOSITORY.createFunction(createFunctionRequest));
+  // TODO: for now, we are not supporting CREATE FUNCTION privilege
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA)
+          """)
+  @AuthorizeKey(METASTORE)
+  public HttpResponse createFunction(@AuthorizeKeys({
+                                        @AuthorizeKey(value = CATALOG, key = "function_info.catalog_name"),
+                                        @AuthorizeKey(value = SCHEMA, key = "function_info.schema_name")
+                                      })
+                                      CreateFunctionRequest createFunctionRequest) {
+    FunctionInfo functionInfo = FUNCTION_REPOSITORY.createFunction(createFunctionRequest);
+    initializeAuthorizations(functionInfo);
+    return HttpResponse.ofJson(functionInfo);
   }
 
   @Get("")
+  @AuthorizeExpression("#defer")
   public HttpResponse listFunctions(
       @Param("catalog_name") String catalogName,
       @Param("schema_name") String schemaName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken) {
-    return HttpResponse.ofJson(
-        FUNCTION_REPOSITORY.listFunctions(catalogName, schemaName, maxResults, pageToken));
+
+    ListFunctionsResponse listFunctionsResponse = FUNCTION_REPOSITORY.listFunctions(catalogName, schemaName, maxResults, pageToken);
+    filterFunctions("""
+            #authorize(#principal, #metastore, OWNER) ||
+            (#authorizeAll(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)) ||
+            (#authorize(#principal, #schema, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #function, OWNER, EXECUTE))
+            """, listFunctionsResponse.getFunctions());
+    return HttpResponse.ofJson(listFunctionsResponse);
   }
 
   @Get("/{name}")
-  public HttpResponse getFunction(@Param("name") String name) {
+  @AuthorizeKey(METASTORE)
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          (#authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #function, OWNER, EXECUTE))
+          """)
+  public HttpResponse getFunction(@Param("name") @AuthorizeKey(FUNCTION) String name) {
     return HttpResponse.ofJson(FUNCTION_REPOSITORY.getFunction(name));
   }
 
   @Delete("/{name}")
+  @AuthorizeKey(METASTORE)
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
```

Comment:
Catalog owners and schema owners who have USE_CATALOG can delete the function. Once those rules are added, this simplifies the next rule.